url7b90c1010b_DastanIqbal_HelloZMQwithJava_tgz-pJ8-ClientJ8/src/zmq/Mtrie.java

```diff
@@ -0,0 +1,394 @@
+
+    //  Add key to the trie. Returns true if it's a new subscription
+    //  rather than a duplicate.
+    public boolean add (byte[] prefix_, int start_, Pipe pipe_)
+    {
+        return add_helper (prefix_, start_, pipe_);
+    }
+
+
+    private boolean add_helper (byte[] prefix_, int start_, Pipe pipe_)
+    {
+        //  We are at the node corresponding to the prefix. We are done.
+        if (prefix_ == null || prefix_.length == start_) {
+            boolean result = pipes == null;
+            if (pipes == null)
+                pipes = new HashSet<Pipe>();
+            pipes.add (pipe_);
+            return result;
+        }
+
+        byte c = prefix_[start_];
+        if (c < min || c >= min + count) {
+
+            //  The character is out of range of currently handled
+            //  charcters. We have to extend the table.
+            if (count == 0) {
+                min = c;
+                count = 1;
+                next = null;
+            }
+            else if (count == 1) {
+                int oldc = min;
+                Mtrie oldp = next[0];
+                count = (min < c ? c - min : min - c) + 1;
+                next = new Mtrie[count];
+                min = Math.min (min, c);
+                next[oldc - min] = oldp;
+            }
+            else if (min < c) {
+
+                //  The new character is above the current character range.
+                count = c - min + 1;
+                next = realloc(next, count, true);
+            }
+            else {
+
+                //  The new character is below the current character range.
+                count = (min + count) - c;
+                next = realloc(next, count, false);
+                min = c;
+            }
+        }
+
+        //  If next node does not exist, create one.
+        if (count == 1) {
+            if (next == null) {
+                next = new Mtrie[1];
+                next[0] = new Mtrie();
+                ++live_nodes;
+                //alloc_assert (next.node);
+            }
+            return next[0].add_helper (prefix_, start_ + 1, pipe_);
+        }
+        else {
+            if (next[c - min] == null) {
+                next[c - min] = new Mtrie();
+                ++live_nodes;
+                //alloc_assert (next.table [c - min]);
+            }
+            return next[c - min].add_helper (prefix_ , start_ + 1, pipe_);
+        }
+    }
+    
+    private Mtrie[] realloc (Mtrie[] table, int size, boolean ended) 
+    {
+        return Utils.realloc(Mtrie.class, table, size, ended);
+    }
+    
+    //  Remove all subscriptions for a specific peer from the trie.
+    //  If there are no subscriptions left on some topics, invoke the
+    //  supplied callback function.
+    public boolean rm (Pipe pipe_, IMtrieHandler func, Object arg) {
+        return rm_helper(pipe_, new byte[0], 0, 0, func, arg );
+    }
+
+    private boolean rm_helper(Pipe pipe_, byte[] buff_, int buffsize_, int maxbuffsize_,
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 20. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 157-171 into a separate method.